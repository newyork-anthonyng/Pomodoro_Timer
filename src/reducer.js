import {
	START_TIMER,
	STOP_TIMER,
	SET_TIME,
	UPDATE_SETTINGS,
	TOGGLE_MODE,
	TOGGLE_SETTINGS_PANEL
} from './actions';

const DEFAULT_WORK_SETTING = 25 * 60;
const DEFAULT_BREAK_SETTING = 5 * 60;

const initialData = {
	isRunning: false,
	seconds: DEFAULT_WORK_SETTING,
	mode: 'work',
	settingsPanelOpen: false,
	default: {
		work: DEFAULT_WORK_SETTING,
		break: DEFAULT_BREAK_SETTING
	}
};

export default function(state = initialData, action) {
	switch(action.type) {
		case START_TIMER:
			console.log('start_timer');
			return Object.assign({}, state, {
				isRunning: !state.isRunning
			});
		case STOP_TIMER:
			console.log('stop_timer');
			return Object.assign({}, state, {
				isRunning: false
			});
		case SET_TIME:
			console.log('setting_time');
			return Object.assign({}, state, {
				seconds: action.seconds
			});
		case UPDATE_SETTINGS:
			console.log('update_settings');
			const newSettings = Object.assign({}, state.default, action.settings);
			return Object.assign({}, state, {
				default: newSettings
			});
		case TOGGLE_MODE:
			console.log('toggling_mode');
			const newMode = state.mode === 'work' ? 'break' : 'work';
			return Object.assign({}, state, {
				mode: newMode
			});
		case TOGGLE_SETTINGS_PANEL:
			console.log('toggle settings panel');
			console.log(state.settingsPanelOpen);
			return Object.assign({}, state, {
				settingsPanelOpen: !state.settingsPanelOpen
			});
		default:
			return state;
	};
};
