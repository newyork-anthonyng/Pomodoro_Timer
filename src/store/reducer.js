import {
  PAUSE_ACTION,
  TOGGLE_ACTION,
  RESET_ACTION,
  SET_TIME_ACTION,
  SET_WORK_INTERVAL_ACTION,
  SET_BREAK_INTERVAL_ACTION
} from "./actions";

export const initialState = {
  mode: "work",
  isPlaying: false,
  time: 25 * 60,
  workInterval: 25 * 60,
  breakInterval: 5 * 60
};

const reducerFunction = (state = initialState, action = {}) => {
  switch (action.type) {
    case PAUSE_ACTION:
      return {
        ...state,
        isPlaying: false
      };
    case TOGGLE_ACTION:
      const userHitsPlayAndTimerIsFinished = state.time === 0 && !state.isPlaying;
      if (userHitsPlayAndTimerIsFinished) {
        const { workInterval, breakInterval } = initialState;
        const newTime = state.mode === "work" ? workInterval : breakInterval;

        return {
          ...state,
          isPlaying: true,
          time: newTime
        };
      }
      return {
        ...state,
        isPlaying: !state.isPlaying
      };
    case RESET_ACTION:
      const { mode, workInterval, breakInterval } = state;
      const newTime = mode === "work" ? workInterval : breakInterval;

      return {
        ...state,
        time: newTime,
        isPlaying: false
      };
    case SET_TIME_ACTION:
      return {
        ...state,
        time: action.data
      };
    case SET_WORK_INTERVAL_ACTION:
      return {
        ...state,
        mode: "work",
        isPlaying: false,
        time: initialState.workInterval
      };
    case SET_BREAK_INTERVAL_ACTION:
      return {
        ...state,
        mode: "break",
        isPlaying: false,
        time: initialState.breakInterval
      };
    default:
        return state;
  }
};

export default reducerFunction;
