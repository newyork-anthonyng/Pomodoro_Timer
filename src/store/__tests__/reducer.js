import reducer, { initialState } from "../reducer";
import {
  PAUSE_ACTION,
  TOGGLE_ACTION,
  RESET_ACTION,
  SET_TIME_ACTION,
  SET_WORK_INTERVAL_ACTION,
  SET_BREAK_INTERVAL_ACTION
} from "../actions";

it("should return initial state by default", () => {
  expect(reducer()).toMatchSnapshot();
});

describe("PAUSE_ACTION", () => {
  it("should return correct state", () => {
    const state = Object.assign({}, initialState, { isPlaying: true });

    expect(reducer(state, { type: PAUSE_ACTION })).toMatchSnapshot();
  });
});

describe("TOGGLE_ACTION", () => {
  let state;

  beforeEach(() => {
    state = Object.assign({}, initialState);
    state.time = 42;
  });

  it("should return correct state when timer was previously playing", () => {
    state.isPlaying = true;

    expect(reducer(state, { type: TOGGLE_ACTION })).toMatchSnapshot();
  });

  it("should return correct state when timer was previously paused", () => {
    state.isPlaying = false;

    expect(reducer(state, { type: TOGGLE_ACTION })).toMatchSnapshot();
  });

  describe("when timer was complete and timer was previously paused", () => {
    beforeEach(() => {
      state.isPlaying = false;
      state.time = 0;
    });

    it("should return correct time when in 'work' mode", () => {
      state.mode = "work";
      expect(reducer(state, { type: TOGGLE_ACTION })).toMatchSnapshot();
    });

    it("should return correct time when in 'break' mode", () => {
      state.mode = "break";
      expect(reducer(state, { type: TOGGLE_ACTION })).toMatchSnapshot();
    });
  });
});

describe("RESET_ACTION", () => {
  let state;

  beforeEach(() => {
    state = Object.assign({}, initialState);
  });

  it("should return correct state in 'work' mode", () => {
    state.mode = "work";

    expect(reducer(state, { type: RESET_ACTION })).toMatchSnapshot();
  });

  it("should return correct state in 'break' mode", () => {
    state.mode = "break";

    expect(reducer(state, { type: RESET_ACTION })).toMatchSnapshot();
  });
});

describe("SET_TIME_ACTION", () => {
  it("should return correct state", () => {
    expect(reducer(initialState, { type: SET_TIME_ACTION, data: 42 })).toMatchSnapshot();
  });
});

describe("SET_WORK_INTERVAL_ACTION", () => {
  it("should return correct state", () => {
    const state = Object.assign({}, initialState);
    state.mode = "break";
    state.isPlaying = true;
    state.time = 42;

    expect(reducer(initialState, { type: SET_WORK_INTERVAL_ACTION })).toMatchSnapshot();
  });
});

describe("SET_BREAK_INTERVAL_ACTION", () => {
  it("should return correct state", () => {
    const state = Object.assign({}, initialState);
    state.mode = "work";
    state.isPlaying = true;
    state.time = 42;

    expect(reducer(initialState, { type: SET_BREAK_INTERVAL_ACTION })).toMatchSnapshot();
  });
});
