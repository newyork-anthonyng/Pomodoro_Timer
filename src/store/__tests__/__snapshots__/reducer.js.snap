// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PAUSE_ACTION should return correct state 1`] = `
Object {
  "breakInterval": 300,
  "isPlaying": false,
  "mode": "work",
  "time": 1500,
  "workInterval": 1500,
}
`;

exports[`RESET_ACTION should return correct state in 'break' mode 1`] = `
Object {
  "breakInterval": 300,
  "isPlaying": false,
  "mode": "break",
  "time": 300,
  "workInterval": 1500,
}
`;

exports[`RESET_ACTION should return correct state in 'work' mode 1`] = `
Object {
  "breakInterval": 300,
  "isPlaying": false,
  "mode": "work",
  "time": 1500,
  "workInterval": 1500,
}
`;

exports[`SET_BREAK_INTERVAL_ACTION should return correct state 1`] = `
Object {
  "breakInterval": 300,
  "isPlaying": false,
  "mode": "break",
  "time": 300,
  "workInterval": 1500,
}
`;

exports[`SET_TIME_ACTION should return correct state 1`] = `
Object {
  "breakInterval": 300,
  "isPlaying": false,
  "mode": "work",
  "time": 42,
  "workInterval": 1500,
}
`;

exports[`SET_WORK_INTERVAL_ACTION should return correct state 1`] = `
Object {
  "breakInterval": 300,
  "isPlaying": false,
  "mode": "work",
  "time": 1500,
  "workInterval": 1500,
}
`;

exports[`TOGGLE_ACTION should return correct state when timer was previously paused 1`] = `
Object {
  "breakInterval": 300,
  "isPlaying": true,
  "mode": "work",
  "time": 42,
  "workInterval": 1500,
}
`;

exports[`TOGGLE_ACTION should return correct state when timer was previously playing 1`] = `
Object {
  "breakInterval": 300,
  "isPlaying": false,
  "mode": "work",
  "time": 42,
  "workInterval": 1500,
}
`;

exports[`TOGGLE_ACTION when timer was complete and timer was previously paused should return correct time when in 'break' mode 1`] = `
Object {
  "breakInterval": 300,
  "isPlaying": true,
  "mode": "break",
  "time": 300,
  "workInterval": 1500,
}
`;

exports[`TOGGLE_ACTION when timer was complete and timer was previously paused should return correct time when in 'work' mode 1`] = `
Object {
  "breakInterval": 300,
  "isPlaying": true,
  "mode": "work",
  "time": 1500,
  "workInterval": 1500,
}
`;

exports[`should return initial state by default 1`] = `
Object {
  "breakInterval": 300,
  "isPlaying": false,
  "mode": "work",
  "time": 1500,
  "workInterval": 1500,
}
`;
